### ISSUE TITLE
Relatively major feature request: keep list of valid categories for error flagging and to avoid duplication

### ISSUE THREAD

**Issue by hschellman** [ivmfnal/metacat#16]
I was running tests and found that asking with runs[any] instead of core.runs[any] gives a python dump instead of an error message.

A very useful feature would be to maintain an internal DB of categories as they are added and then, if one gets this kind of error catch it and suggest that the person check that their category is in the list.

Not being able to generate a list of already defined parameters in sam has always been a pain.  I think we could catch them as they go in and then be able to either ask the user to check the list or do it ourselves when we get a syntax error. 

End users would love to have a list of the categories, even better would be if they were required to describe them.

Here is my example - my screwup but the error message was extremely uninformative. 
<schellma>  metacat query -i  "files  from dune:all where core.file_type=detector and core.run_type='protodune-sp' and core.data_stream=physics and runs[any] in 5141:5200" > physics_ids.txt
Traceback (most recent call last):
  File "/nashome/s/schellma/.local/bin/metacat", line 33, in <module>
    sys.exit(load_entry_point('metacat==3.16.1', 'console_scripts', 'metacat')())
  File "/cvmfs/dune.opensciencegrid.org/products/dune/metacat/v3_9_7/NULL/lib/python3/site-packages/metacat/ui/metacat_ui.py", line 91, in main
    cli.run(sys.argv, argv0="metacat")
  File "/cvmfs/dune.opensciencegrid.org/products/dune/metacat/v3_9_7/NULL/lib/python3/site-packages/metacat/ui/cli/cli.py", line 205, in run
    self._run(command, context, argv, usage_on_error)
  File "/cvmfs/dune.opensciencegrid.org/products/dune/metacat/v3_9_7/NULL/lib/python3/site-packages/metacat/ui/cli/cli.py", line 200, in _run
    return interp._run(pre_command + word, context, rest, usage_on_error = usage_on_error)
  File "/cvmfs/dune.opensciencegrid.org/products/dune/metacat/v3_9_7/NULL/lib/python3/site-packages/metacat/ui/cli/cli.py", line 99, in _run
    return self(command, context, opts, args)
  File "/cvmfs/dune.opensciencegrid.org/products/dune/metacat/v3_9_7/NULL/lib/python3/site-packages/metacat/ui/metacat_query.py", line 118, in __call__
    print("%s%s" % (f["fid"],meta_out))
TypeError: string indices must be integers
<schellma>  metacat query -i  "files  from du

**Comment by ivmfnal** [comment]
If I run a query like that I get this error instead of the dump:

```
$ ui/metacat query files from dc4:dc4 where "runs[any] = 32"
Server side error: MQLExecutionError: File attribute runs value must be a scalar. Got array_any instead
```

I see that you run very old version of the client. Current versionis 1.17.3:

```
$ pip show metacat
Name: metacat
Version: 3.17.3
Summary: MetaCat is a general purpose metadata database. This package is the client side portion of the product.
Home-page: https://github.com/ivmfnal/metacat
Author: Igor Mandrichenko
Author-email: ivm@fnal.gov
License: BSD 3-clause
Location: /Users/ivm/miniconda3/lib/python3.7/site-packages
Requires: pythreader, pyjwt, requests
Required-by: datadispatcher
```

As for the list of categories, they are kept and viewable via GUI: https://metacat.fnal.gov:9443/dune_meta_demo/app/gui/categories/index

**Comment by ivmfnal** [comment]
Definition of category parameters is something an admin (or the collaboration) has to do. Current mechanism is via GUI.

**Comment by hschellman** [comment]
Yes but can we maintain a DB of them that one can query?  Ie

metacat categories list

gives

core.runs  array ownership description

or

metacat query “categories from “namespace” …."

or something like that.

Not having that in sam was a disaster.   If people knew how to find categories they would not create so many and the UI could check for validity.




On Oct 12, 2022, at 9:40 AM, Igor Mandrichenko ***@***.******@***.***>> wrote:


[This email originated from outside of OSU. Use caution with links and attachments.]


Definition of category parameters is something an admin (or the collaboration) has to do. Current mechanism is via GUI.

**Comment by ivmfnal** [comment]
Basic support for category listing and displaying is added to CLI and API. Please upgrade the client to 3.19.0.
If additional functionality is needed, let's open a new issue.
